test/compiler.test.js

```diff
@@ -28,8 +28,34 @@ describe('compiler', function() {
         models: {
           'foo-bar-bat-baz': {
             dataSource: 'the-db'
+          },
+          'foo-bar-bat-baz-with-def': {
+            dataSource: 'the-db'
+          },
+          'foo-bar-bat-baz-with-def-no-source': {
+            dataSource: 'the-db'
           }
         },
+        modelDefinitions: [
+          {
+            definition: {
+              name: 'foo-bar-bat-baz-with-def'
+            },
+            sourceFile: path.join(SIMPLE_APP, 'common', 'models', 'foo-bar-bat-baz-with-def.js')
+          },
+          {
+            definition: { // sourceFile is a not existing file.
+              name: 'foo-bar-bat-baz-with-def-no-source'
+            },
+            sourceFile: path.join(SIMPLE_APP, 'common', 'models', 'foo-bar-bat-baz-with-def!!!.js')
+          },
+          {
+            definition: {
+              name: 'i-should-not-exist' //not in 'models'
+            },
+            sourceFile: path.join(SIMPLE_APP, 'common', 'models', 'i-should-not-exist.js')
+          }
+        ],
```

Comment:
This "from options" block is already pretty bad, please don't make it worse.

Write a new standalone test for each different scenario, see the following tests for inspiration: https://github.com/shlomiassaf/loopback-boot/blob/d9d0e0297bf8007649778e25729d4b7d8077921c/test/compiler.test.js#L580-L723